### PR TITLE
[Enhancement] Implement Distance-Based Detection Matching in DetectionMetrics

### PR DESCRIPTION
**Pull request by DimaBir**
## Pull Request Description

This pull request introduces enhancements and additions related to distance-based metrics in the `DetectionMetrics` class. The main changes include:

- **Addition**: The `DetectionMetricsDistanceBased` class has been added to support distance-based metrics for object detection.
 Abstract base class `DetectionMatching`, which defines the interface for matching implementations. 
- **Modifications**: The existing `DetectionMetrics` class has been extended to incorporate distance-based metrics when calculating precision, recall, F1, and mAP.

These changes improve the capabilities of the metrics calculation for object detection tasks, allowing users to analyze detection performance using distance-based criteria. Below are the details of the changes made:

## Changes Made

- Introduced the `DetectionMetricsDistanceBased` class, which extends the existing `DetectionMetrics` class to support distance-based metrics.
- Modified the `DetectionMetrics` class to accept distance metric definitions and thresholds.
- Updated the `compute` method in both classes to calculate metrics based on the results of distance-based matching when using `DetectionMetricsDistanceBased.`
- Introduced an abstract base class, `DetectionMatching`, which defines the interface for matching implementations. It includes abstract methods for getting thresholds and computing metrics for targets and crowd targets.
- Implemented the `IoUMatching` class to maintain backward compatibility with IoU-based metrics.
- Added the `DistanceMatching` class for computing distance-based matching between predictions and targets.

## Proposed Changes

- The changes have been tested to ensure the accuracy and reliability of the distance-based metrics calculations.
- The new classes and methods maintain compatibility with the existing functionality, ensuring backward compatibility.

## Testing

- Extensive testing has been performed to validate the accuracy and correctness of the metrics calculation.
- The test cases cover various scenarios and edge cases for distance-based metrics.

This PR is ready for review and feedback. I would greatly appreciate your suggestions and suggestions for improvement.

### CODE REVIEW

**Comment by BloodAxe** [comment]
Thanks for your PR. It looks like your PR contains commits that are not signed. We required all commits in the PR to be signed. You can use this suggestion to address that: https://github.com/Deci-AI/super-gradients/pull/1465#issuecomment-1729839476

Maybe I'm missing something, but the whole idea of the PR is to allow using some other metric when matching pred vs gt boxes. So I have few questions regarding this:
1) What are the use cases when one may want to use other metric?
2) Can you provide any papers / project using custom bbox similarity metric?

I'm just not convinced this feature would be relevant for many users.

**Comment by DimaBir** [comment]
> Thanks for your PR. It looks like your PR contains commits that are not signed. We required all commits in the PR to be signed. You can use this suggestion to address that: [#1465 (comment)](https://github.com/Deci-AI/super-gradients/pull/1465#issuecomment-1729839476)
> 
> Maybe I'm missing something, but the whole idea of the PR is to allow using some other metric when matching pred vs gt boxes. So I have few questions regarding this:
> 
> 1. What are the use cases when one may want to use other metric?
> 2. Can you provide any papers / project using custom bbox similarity metric?
> 
> I'm just not convinced this feature would be relevant for many users.

Thank you for your feedback, and I apologize for the oversight regarding the signed commits. I will address this as soon as possible by signing all the commits in the PR according to the repository's guidelines. I will also review the suggestions in [#1465 (comment)](https://github.com/Deci-AI/super-gradients/pull/1463#issuecomment-926123456) to ensure compliance.

Regarding your queries about the use of different metrics for bounding box similarity, here are my thoughts:

### 1. Use Cases for Different Metrics:
 - **Distance-Based Metric:**
    The task was to implement the distance-based metric to enhance existing Detection Metric to allow metrics based on distances from the bboxes' centers, particularly for small bboxes less than 5 pixels.
- **Optimization for Specific Criteria:**
     Custom metrics can be optimized for specific criteria, such as maximizing the overlap or minimizing the distance between bounding boxes, allowing for more fine-tuned evaluations.
  
### 2. References to Papers/Projects:
   While I don't have a specific paper or project that uses a custom bbox similarity metric, the flexibility to use different metrics can be crucial for research and development in object detection, allowing researchers to explore and evaluate different approaches and methodologies.

### 3. Relevance of the Feature:
   While the relevance of this feature might take time to become apparent, providing the flexibility to use different metrics can cater to a wider range of use cases and user needs.

**Comment by BloodAxe** [comment]
@DimaBir I would like to proceed with this PR and get it merged. Can you please fix the merge conflict?

**Comment by DimaBir** [comment]
> @DimaBir I would like to proceed with this PR and get it merged. Can you please fix the merge conflict?

Hi @BloodAxe, sure!
I see that merge already started. IMHO, it should be fine.
If there are still some conflicts, I will fix them until EoD.

**Comment by BloodAxe** [comment]
Awesome, thanks!

File: src/super_gradients/training/utils/detection_utils.py [hunk 1]
```diff
@@ -752,6 +752,199 @@ def crowd_ioa(det_box: torch.Tensor, crowd_box: torch.Tensor) -> torch.Tensor:
     return inter / det_area[:, None]  # crowd_ioa = inter / det_area
 
 
+class DetectionMatching(ABC):
```
Comment (BloodAxe) [inline src/super_gradients/training/utils/detection_utils.py:755]: Please add missing docstring to this class (brief description of the class goal) and to each abstract method (Specifically what are the expected return types are)

**Comment by DimaBir** [comment]
Hi @BloodAxe, thanks for pointing this out. I will do np, tomorrow is it okay? 

I have a few meetings to attend today 🫤

File: tests/unit_tests/detection_metrics_distance_based_test.py [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+import unittest
+import torch
+import random
+from typing import List, Optional, Tuple
+
+from super_gradients.training.metrics.detection_metrics import DetectionMetricsDistanceBased
+from super_gradients.training.utils.detection_utils import EuclideanDistance, ManhattanDistance
+
+
+class TestDetectionMetricsDistanceBased(unittest.TestCase):
```
Comment (BloodAxe) [inline tests/unit_tests/detection_metrics_distance_based_test.py:11]: The test case is not registered in tests/deci_core_unit_test_suite_runner.py

**Comment by BloodAxe** [comment]
@DimaBir thanks for updating PR. Assuming build passes the overall PR looks good to me except a few inline comments I left. Nothing serious, mostly docs improvement.
However there is one blocker thing that needs to be addressed - this PR still contains unsigned commits and for that reason CI will not allow merging it. Fortunately the fix is pretty simple: https://github.com/Deci-AI/super-gradients/blob/master/CONTRIBUTING.md#i-have-created-pr-with-unsigned-commits-do-i-have-to-start-over

**Comment by DimaBir** [comment]
Sure @BloodAxe I will fix everything and fix commits to signed.

File: src/super_gradients/training/utils/detection_utils.py (hunk 1 above)
Comment (DimaBir) [reply src/super_gradients/training/utils/detection_utils.py:755]: Added

File: src/super_gradients/training/utils/detection_utils.py [hunk 3]
```diff
@@ -752,6 +752,199 @@ def crowd_ioa(det_box: torch.Tensor, crowd_box: torch.Tensor) -> torch.Tensor:
...
+    @abstractmethod
+    def get_thresholds(self):
```
[the remark this answers is not in the document]
Comment (DimaBir) [reply src/super_gradients/training/utils/detection_utils.py]: Added annotations

**Comment by DimaBir** [comment]
> @DimaBir thanks for updating PR. Assuming build passes the overall PR looks good to me except a few inline comments I left. Nothing serious, mostly docs improvement. However there is one blocker thing that needs to be addressed - this PR still contains unsigned commits and for that reason CI will not allow merging it. Fortunately the fix is pretty simple: https://github.com/Deci-AI/super-gradients/blob/master/CONTRIBUTING.md#i-have-created-pr-with-unsigned-commits-do-i-have-to-start-over

Done, force pushed branch with signed commits

**Review by BloodAxe — APPROVED** [review]
LGTM
